### ISSUE TITLE
Improve Tasks

### ISSUE THREAD

**Issue by alajmo** [alajmo/sake#16]
This is an issue to improve the **task** object. I've gathered the following use-cases:

- to omit certain tasks from auto-completion and/or being called directly (mainly tasks which are called by other tasks)
- to repress certain task output if exit code is 0, otherwise displayed
- a summary of task execution at the end
- to be able to pass environment variables between tasks
- to be able to access the previous exit code of the previous task
- conditional task execution
  - tags/servers filtering
  - ensure command (check existence of file, software, etc.)
- on-error/on-success task execution
- a cleanup task that is always ran
- to log task execution to a file
- to abort if certain env variables are not present (required envs)
- want a --step mode flag or config setting to prompt before executing a task

A separate pull-request will be created to implement the necessary changes.

**Comment by eppixx** [comment]
Currently it is possible to have duplicate task without error.

**Comment by alajmo** [comment]
Yes, I had plans to rectify that, however, it makes sense to keep it as it is since you might want to override tasks. 

So it would be possible to have a task defined in your home config (`~/.config/sake/config.yaml`) and then override that task if you're working with a config that defines the same task.

**Comment by eppixx** [comment]
I get that point. But then there should be a way to see which one is overwritten (maybe a flag to check the config).
The worst case scenario is when you have a sprawling config and somebody edits the config and nothing changes because it is overwritten somewhere else.

**Comment by alajmo** [comment]
> I get that point. But then there should be a way to see which one is overwritten (maybe a flag to check the config). The worst case scenario is when you have a sprawling config and somebody edits the config and nothing changes because it is overwritten somewhere else.

I've had some second thoughts here, initially, I thought the use case was valid (users sharing the same code base could override tasks), however, I don't see that use case happening a lot, and that it should take precedence over debugability or worst case, you think you're running a task, but it's some other task you defined in your user config.
Will add it to the roadmap.
